src/main/java/spoon/support/adaption/TypeAdaptor.java

```diff
@@ -0,0 +1,598 @@
+	 */
+	public TypeAdaptor(CtType<?> hierarchyStart) {
+		this.hierarchyStart = hierarchyStart;
+		this.hierarchyStartReference = hierarchyStart.getReference();
+		this.initializedWithReference = false;
+	}
+
+	/**
+	 * Creates a new type adaptor using the given reference as the start of its hierarchy.
+	 *
+	 * @param hierarchyStart the start of the hierarchy
+	 */
+	public TypeAdaptor(CtTypeReference<?> hierarchyStart) {
+		CtTypeReference<?> usedHierarchyStart = hierarchyStart;
+		if (hierarchyStart instanceof CtArrayTypeReference) {
+			usedHierarchyStart = ((CtArrayTypeReference<?>) hierarchyStart).getArrayType();
+		}
+		this.hierarchyStartReference = usedHierarchyStart;
+		this.hierarchyStart = hierarchyStartReference.getTypeDeclaration();
+		this.initializedWithReference = true;
+	}
+
+	/**
+	 * Creates a new type adaptor using the given method as the start of its hierarchy.
+	 *
+	 * @param hierarchyStart the start of the hierarchy
+	 */
+	public TypeAdaptor(CtMethod<?> hierarchyStart) {
+		this(hierarchyStart.getDeclaringType());
+		this.startMethod = hierarchyStart;
+	}
+
+	/**
+	 * Checks if the context of this type adapter is a subtype of the passed superRef.
+	 *
+	 * @param superRef the super reference to check against
+	 * @return true if the context of this type adapter is a subtype of the passed superRef
+	 * @implNote This implementation behaves the same as {@code isSubtype(hierarchyStart,
+	 * 	superRef)}
+	 * @see #isSubtype(CtType, CtTypeReference)
+	 */
+	@SuppressWarnings("removal")
+	public boolean isSubtypeOf(CtTypeReference<?> superRef) {
+		if (superRef.getFactory().getEnvironment().useOldAndSoonDeprecatedClassContextTypeAdaption()) {
+			return getOldClassTypingContext().isSubtypeOf(superRef);
+		}
+		// We have no declaration so we can't really do any subtype queries
+		if (hierarchyStart == null) {
+			return false;
+		}
+
+		boolean subtype = isSubtype(hierarchyStart, superRef);
+		if (!subtype) {
+			return false;
+		}
+		if (hierarchyStartReference.getActualTypeArguments().isEmpty() && superRef.getActualTypeArguments().isEmpty()) {
+			return true;
+		}
+		return new ClassTypingContext(hierarchyStartReference).isSubtypeOf(superRef);
+	}
+
+	/**
+	 * @return the context of this type adaptor
+	 */
+	public CtType<?> getHierarchyStart() {
+		return hierarchyStart;
+	}
+
+	/**
+	 * Checks whether the base is a subtype of the passed superref. Generic parameters in superRef are
+	 * ignored.
+	 *
+	 * @param base the base type
+	 * @param superRef the potential supertype
+	 * @return true if base extends/implements the super type
+	 */
+	@SuppressWarnings("removal")
+	public static boolean isSubtype(CtType<?> base, CtTypeReference<?> superRef) {
+		if (base.getFactory().getEnvironment().useOldAndSoonDeprecatedClassContextTypeAdaption()) {
+			return new TypeAdaptor(base).isSubtypeOf(superRef);
+		}
+		String superRefFqn = superRef.getTypeErasure().getQualifiedName();
+
+		// Everything inherits from object
```

Comment:
```suggestion
```

Reply:
I would say these comments are superfluous, but if you really want to keep them, put them _inside_ the if-statement they are related to. That connects them more strongly to the statement they relate to.

Reply:
I have deleted a few and moved a few more into ifs. I hope the ones still left have enough merit to earn their stay.